File: lib/plugins/opabsl/mlbsl/bslTime.ml
(*
    Copyright © 2011, 2012 MLstate

    This file is part of Opa.

    Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*)
(** Warning : the slicer and the serialization / unserialization is not ready
    to have abstract types for values which needs to be passed client <--> server
    So, the type time cannot be abstract here.

    Todo : json handling in libbsl *)

   ##extern-type time_t = int

   let unwrap = Time.in_milliseconds
   let wrap = Time.milliseconds

   ##register now : -> time_t
   let now () = unwrap (Time.now ())

   ##register gmt_msec : time_t -> int
   let gmt_msec t = Time.gmt_msec (wrap t)

   ##register gmt_sec : time_t -> int
   let gmt_sec t = Time.gmt_sec (wrap t)

   ##register gmt_min : time_t -> int
   let gmt_min t = Time.gmt_min (wrap t)

   ##register gmt_hour : time_t -> int
   let gmt_hour t = Time.gmt_hour (wrap t)

   ##register gmt_mday : time_t -> int
   let gmt_mday t = Time.gmt_mday (wrap t)

   ##register gmt_mon : time_t -> int
   let gmt_mon t = Time.gmt_mon (wrap t)

   ##register gmt_year : time_t -> int
   let gmt_year t = Time.gmt_year (wrap t)

   ##register gmt_wday : time_t -> int
   let gmt_wday t = Time.gmt_wday (wrap t)

   ##register gmt_yday : time_t -> int
   let gmt_yday t = Time.gmt_yday (wrap t)

   ##register gmt_isdst : time_t -> bool
   let gmt_isdst t = Time.gmt_isdst (wrap t)

   ##register local_msec : time_t -> int
   let local_msec t = Time.local_msec (wrap t)

   ##register local_sec : time_t -> int
   let local_sec t = Time.local_sec (wrap t)

   ##register local_min : time_t -> int
   let local_min t = Time.local_min (wrap t)

   ##register local_hour : time_t -> int
   let local_hour t = Time.local_hour (wrap t)

   ##register local_timezone_offset : -> int
   let local_timezone_offset _ = Time.local_timezone_offset()

   ##register local_mday : time_t -> int
   let local_mday t = Time.local_mday (wrap t)

   ##register local_mon : time_t -> int
   let local_mon t = Time.local_mon (wrap t)

   ##register local_year : time_t -> int
   let local_year t = Time.local_year (wrap t)

   ##register local_wday : time_t -> int
   let local_wday t = Time.local_wday (wrap t)

   ##register local_yday : time_t -> int
   let local_yday t = Time.local_yday (wrap t)

   ##register local_isdst : time_t -> bool
   let local_isdst t = Time.local_isdst (wrap t)

   ##register mktime : int, int, int, int, int, int, int -> time_t
   let mktime year month day h min sec ms = unwrap (Time.mktime ~year ~month ~day ~h ~min ~sec ~ms)

   ##register import_t : time_t -> int
   let import_t t = t

   ##register export_t : int -> time_t
   let export_t t = t

   ##register get_accurate_time : -> float
   let get_accurate_time = Time.get_accurate_time

File: lib/plugins/opabsl/jsbsl/bslTime.js
/*
    Copyright © 2011, 2012 MLstate

    This file is part of Opa.

    Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

/**
 * Time primitives
 * @side{both}
 *
 * BEWARE : Major of bypasses which manipulates time_t are tagged as pure
 * because we define only getter.
 */

var time_month_duration = new Array(31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31);

function leap_year(y){
    return( ((y % 4) == 0) && ((y % 100) != 0) || ((y % 400) == 0) );
}

/**
 * Convert an integer to a string representation with at least two digits.
 *
 * @param {number} n A non-negative number.
 * @return {string} Either "0" followed by the string representation of [n] (if n < 10) or simply the string representation of [n] (otherwise)
 */
function two_digits(n)
{
    if (n < 10)
        return "0" + n;
    else
        return n.toString();
}

/**
 * Return the current time, as an [int]
 *
 * @return {number} The current time, as a delta from the epoch.
 * @register { -> time_t}
 */
function now() {
    var t = new Date();
    return t.getTime();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function gmt_msec(f) {
    var t = new Date(); t.setTime(f);
    return t.getUTCMilliseconds();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function gmt_sec(f) {
   var t = new Date(); t.setTime(f);
    return t.getUTCSeconds();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function gmt_min(f) {
   var t = new Date(); t.setTime(f);
    return t.getUTCMinutes();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function gmt_hour(f) {
    var t = new Date(); t.setTime(f);
    return t.getUTCHours();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function gmt_mday(f) {
    var t = new Date(); t.setTime(f);
    return t.getUTCDate();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function gmt_mon(f) {
    var t = new Date(); t.setTime(f);
    return t.getUTCMonth();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function gmt_year(f) {
    var t = new Date(); t.setTime(f);
    return t.getUTCFullYear();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function gmt_wday(f) {
    var t = new Date(); t.setTime(f);
    return t.getUTCDay();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function gmt_yday(f) {
    var t = new Date(); t.setTime(f);
    var yday = t.getUTCDate() - 1;
    var mon = t.getUTCMonth();
    for (var m = 0 ; m < mon ; m++)
        yday += time_month_duration[m];
    if (mon > 1 && leap_year(t.getUTCFullYear()))
        yday++;
    return yday;
}

/**
 * @register {time_t -> bool}
 * @pure
 */
function gmt_isdst(f) {
    var t = new Date(); t.setTime(f);
    return false;
}

/**
 * @register {time_t -> int}
 * @pure
 */
function local_msec(f) {
    var t = new Date(); t.setTime(f);
    return t.getMilliseconds();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function local_sec(f) {
    var t = new Date(); t.setTime(f);
    return t.getSeconds();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function local_min(f) {
    var t = new Date(); t.setTime(f);
    return t.getMinutes();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function local_hour(f) {
    var t = new Date(); t.setTime(f);
    return t.getHours();
}

/**
 * @register {-> int}
 * @pure
 */
function local_timezone_offset() {
    var t = new Date();
    return t.getTimezoneOffset();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function local_mday(f) {
    var t = new Date(); t.setTime(f);
    return t.getDate();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function local_mon(f) {
    var t = new Date(); t.setTime(f);
    return t.getMonth();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function local_year(f) {
    var t = new Date(); t.setTime(f);
    return t.getFullYear();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function local_wday(f) {
    var t = new Date(); t.setTime(f);
    return t.getDay();
}

/**
 * @register {time_t -> int}
 * @pure
 */
function local_yday(f) {
    var t = new Date(); t.setTime(f);
    var yday = t.getDate() - 1;
    var mon = t.getMonth();
    for (var m = 0 ; m < mon ; m++)
        yday += time_month_duration[m];
    if (mon > 1 && leap_year(t.getFullYear()))
        yday++;
    return yday;
}

/**
 * @register {time_t -> bool}
 * @pure
 */
function local_isdst(f) {
    var t = new Date(); t.setTime(f);
    return false;
}

/**
 * @register {int,int,int,int,int,int,int -> time_t}
 * @pure
 */
function mktime(year, mon, day, hour, min, sec, ms) {
    var t = new Date();
    t.setFullYear(year, mon, day);
    t.setHours(hour, min, sec);
    t.setMilliseconds(ms);
    return (t.getTime());
}

/**
 * @register {time_t -> int}
 * @pure
 */
function import_t(t) {
    return t;
}

/**
 * @register {int -> time_t}
 * @pure
 */
function export_t(t) {
    return t;
}

/** @register { -> float} get_accurate_time get_accurate_time */
function get_accurate_time()
{
    var t = new Date();
    return t.getTime()/1000.0;
}
